File: kometa-cat.js
#!/usr/bin/env node
// kometa-cat.js
// Visualize homoglyph carriers in text.
// Highlights Cyrillic (red), Greek (green), Latin carriers (blue).

"use strict";
const fs = require("fs");

// Character sets from kometa.js
const ALPHA = {
  lat: "KOMETAXPHo",
  cyr: "КОМЕТАХРНо",
  ell: "ΚΟΜΕТΑΧΡΗο",
};

const BETA = {
  lat: "IJij",
  cyr: "ІЈіј",
};

// Build lookup: character → script type
const charToScript = new Map();
const addToMap = (script, chars) => {
  for (const ch of chars) charToScript.set(ch, script);
};

Object.entries(ALPHA).forEach(([script, chars]) => addToMap(script, chars));
Object.entries(BETA).forEach(([script, chars]) => addToMap(script, chars));

// ANSI color codes
const COLORS = {
  lat: "\x1b[34m", // Blue - Latin carrier (inactive)
  cyr: "\x1b[31m", // Red - Cyrillic (bit 0)
  ell: "\x1b[32m", // Green - Greek (bit 1)
  reset: "\x1b[0m",
};

// Colorize entire text
const colorize = text => {
  return [...text]
    .map(ch => {
      const script = charToScript.get(ch);
      return script ? COLORS[script] + ch + COLORS.reset : ch;
    })
    .join("");
};

// ── Main ─────────────────────────────────────
const [, , file] = process.argv;

if (file) {
  const content = fs.readFileSync(file, "utf8");
  process.stdout.write(colorize(content));
} else {
  let data = "";
  process.stdin.on("data", chunk => {
    data += chunk;
  });
  process.stdin.on("end", () => {
    process.stdout.write(colorize(data));
  });
}

File: kometa-cat.py
#!/usr/bin/env python3
# kometa-cat.py
# Visualise homoglyph carriers in text.
# Highlights Cyrillic (red), Greek (green), Latin carriers (blue).
# Imports ALPHA directly from kometa.py — single source of truth.

import sys, os
sys.path.insert(0, os.path.dirname(__file__))
from kometa import DICT

COLORS = dict(lat="\x1b[34m", cyr="\x1b[31m", reset="\x1b[0m")

LOOKUP = {}
for _i, (_l, _c) in enumerate(zip(DICT["lat"], DICT["cyr"])):
    LOOKUP[_l] = "lat"
    LOOKUP[_c] = "cyr"

def _summary(text: str) -> str:
    counts = {"lat": 0, "cyr": 0}
    for ch in text:
        if ch in LOOKUP: counts[LOOKUP[ch]] += 1
    total = sum(counts.values())
    if not total: return "no carriers found"
    return (f"{COLORS['cyr']}cyr:{counts['cyr']}{COLORS['reset']}  "
            f"{COLORS['lat']}lat:{counts['lat']}{COLORS['reset']}  "
            f"total:{total}")

def colorize(text: str) -> str:
    out = []
    for ch in text:
        script = LOOKUP.get(ch)
        out.append(f"{COLORS[script]}{ch}{COLORS['reset']}" if script else ch)
    return "".join(out)

if __name__ == "__main__":
    if len(sys.argv) > 1:
        text = open(sys.argv[1], encoding="utf-8").read()
    else:
        text = sys.stdin.read()
    sys.stdout.write(colorize(text) + "\n")
    sys.stderr.write(_summary(text) + "\n")
